File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def __init__(self):
        self.ans = True

    def dfs(self, root):

        if root is None:
            return 0

        
        L = self.dfs(root.left)
        R = self.dfs(root.right)
        if abs(L - R) > 1:
            self.ans = False

        return 1 + max(L, R)

    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        if root is None:
            return True

        self.ans = True

        self.dfs(root)

        return self.ans

        

File: Data Structures & Algorithms/house-robber-iii/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int, int> dfs(TreeNode* root){
        if(!root){
            return {0, 0};
        }

        auto [LRob, LNotRob] = dfs(root->left);
        auto [RRob, RNotRob] = dfs(root->right);

        int robThis = root->val + LNotRob + RNotRob;

        int skipThis = max(LRob, LNotRob) + max(RRob, RNotRob);

        return {robThis, skipThis};
    }
    

    int rob(TreeNode* root) {
        auto [rob, notRob] = dfs(root);

        return max(rob, notRob);
    }
};